---
title: "R Workshop"
subtitle: "10/04/2019"
output:
  html_document:
    theme: cosmo
    toc: yes 
---

```{r global, echo=F}
knitr::opts_chunk$set(message=FALSE, warning=FALSE, eval=FALSE, cache=F)
```

# Intro

For today's workshop, we're going to use R to go through a typical bioinformatics analysis workflow. We're going to use common bioinformatics techniques to pre-process and analyze gene expression data related to breast cancer. 

The data we will analyze is breast cancer RNA-Seq data from [TCGA](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga), a popular publicly-available database for cancer-related datasets. The goal of the analysis will be to identify genes that show significant changes in expression between normal and tumor tissues, followed by identifying the pathways they are associated with. After importing the data and performing some data pre-processing, we will carry out differntial expression analysis and gene set enrichment analysis.

Main steps in today's workshop:

1. Import and pre-process RNA-Seq data
2. Identify differentially-expressed genes between tumor and control samples
3. Identify significantly-enriched pathways in the gene sets


Make sure to have the following packages installed for this workshop:

* `Biobase`
* `dplyr`
* `DESeq2`
* `fgsea`
* `ggplot2`
* `msigdbr`
* `fgsea`

# Working with Expression Set Objects
An expression set is a data object consisting of three entities: the
expression matrix (`exprs`), the phenotye data (`pData`), and the
feature data (`fData`).

![](media/eset.png)

We read in the RDS file included in this repo. It corresponds to a subset of samples from a gene expression dataset of breast cancer (BRCA) primary tissue samples from the TCGA project.

```{r, eval=T}
library(Biobase)
library(magrittr)
library(dplyr)
library(ggplot2)
library(DESeq2)
library(fgsea)
```

```{r readdata, eval=T}
brca <- readRDS("data/TCGA-BRCA.rds")

# dimensions of the expression data
dim(brca)

# dimensions of the gene annotation
dim(fData(brca))
# first few rows of gene annotations
head(fData(brca)[,c("ensembl_transcript_id", "ensembl_gene_id", "hgnc_symbol")])

# dimensions of the phenotypic annotation
dim(pData(brca))
# first few rows of phenotype
head(pData(brca)[,c("patient_id", "sample_type", "tumor_subtype")])

# how many of each sample type?
table(pData(brca)$sample_type)

# how many tumor subtypes?
table(pData(brca)$tumor_subtype)
```

One of the advantages of using an ExpressionSet is that the three component objects are always properly paired, and subsetting can be carried out straightforwardly.
```{r subsetting, eval=T}
#subset first 100 genes and 10 samples
brcaSubset <- brca[1:100,1:10]

#updated dimensions after subsetting
dim(brcaSubset)
dim(pData(brcaSubset))
dim(fData(brcaSubset))
```

# Data Wrangling

This section of the workshop will look at some of the common data pre-processing steps.

### Pre-processing
Do the following:  
1. Extract the phenotype dataframe from the expression set  
2. Filter out "Metastatic" tumors  
3. Keep only "Solid Tissue Normal" samples and "Primary Tumor" samples subtyped as "Basal"  
4. Rename entries in the "sample_type" column as "tumor" and "normal"  
5. Subset the original expression set by the filtered samples  
6. Replace the pData slot with the updated data frame  

```{r echo=F, eval=T}
pdat <- pData(brca) %>%
        filter(sample_type != "Metastatic") %>%
        filter(sample_type == "Solid Tissue Normal" | tumor_subtype == "Basal") %>%
        mutate(sample_type=ifelse(sample_type == "Primary Tumor", "tumor", "normal")) %>%
        set_rownames(.[,"full_id"])

brca.sub <- brca[,rownames(pdat)]
pData(brca.sub) <- pdat
```

```{r}
brca.sub <- ...
```


Your output should have the following dimensions:

```{r, eval=T}
dim(brca.sub)
table(brca.sub$sample_type)
table(brca.sub$tumor_subtype)
```

### Variation Filtering
Filter out genes that have zero variance across samples.
```{r echo=F, eval=T}
brca.filtered <- brca.sub[apply(exprs(brca.sub), 1, var) != 0,]
```

```{r}
brca.filtered <- ...
```

Your output should have the following dimensions:

```{r, eval=T}
dim(brca.filtered)
```

### Low-Counts Filtering
Filter out genes that have zero counts in at least 1/5 of samples.
```{r echo=F, eval=T}
brca.filtered <- brca.filtered[apply(exprs(brca.filtered) != 0, 1, sum)/ncol(brca.filtered) >= 1/5,]
```

```{r}
brca.filtered <- ...
```

Your output should have the following dimensions:


```{r, eval=T}
dim(brca.filtered)
```

### Log-transformation

Log-transform the counts. (Hint: what is log2(0)?)
```{r, eval=T, echo=F}
brca.filtered.log <- brca.filtered
exprs(brca.filtered.log) <- log2(exprs(brca.filtered) + 1) # Calculate log2 counts
```

```{r}
brca.filtered.log <- ...
```


Show first 5 rows and 2 samples after log-transformation:
```{r, eval=T}
exprs(brca.filtered.log)[1:5,1:2]
```

# Differential Expression Analysis

Differential expression analysis aims to find genes that show a significant change in expression levels between two conditions. Some of the commonly-used R packages for this analysis include *DESeq2*, *edgeR*, and *limma*. For this workshop, we will use *DESeq2* to identify genes that are differentially-expressed between normal and tumor tissue in our dataset. The [DESeq2 vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) offers an in-depth tutorial that could be helpful for completing this section of the workshop.

### Set up DESeq2 object

The first step is to create a `DESeqDataSet` object. This object includes a variable `design` which should be set to the variables that will be used in the model (e.g. variable to compare, confounding factors). Note that DESeq2 takes the raw counts matrix (un-normalized, not log-transformed) as input, so use the `brca.filtered` object from the previous section..

Hint: use the `DESeqDataSetFromMatrix` function to set up this object using our filtered counts matrix.

```{r deseq2object, eval=T, echo=F}
# Subset first 500 genes and 100 samples for faster run-time
brca.subset <- brca.filtered[1:500,1:100]

#import package
library(DESeq2)

# Create DESeq2 specific object
dds <- DESeqDataSetFromMatrix(countData = exprs(brca.subset),
                              colData = pData(brca.subset),
                              design = formula(~sample_type)) 
```

```{r}
#import package
library(DESeq2)

# Create DESeq2 specific object
dds <- ...
```

### Relevel factors

Before we continue, we should set our normal tissue samples as our reference level for comparison in our differential expression analysis.

Hint: use the `relevel` function.

```{r relevelFactor, eval=T, echo=F}
# Set reference level to 'normal'
dds@colData$sample_type <- relevel(dds@colData$sample_type, ref = "normal")
```

```{r}
dds@colData$sample_type <- ...
```

### Run DESeq2

Now we can perform the differential expression test and store the results as a table. The results show us the mean expression of the gene, the log fold change in expression between the two sample types, along with p-values and adjusted p-values (FDR) for each gene.

```{r}
# Run DESeq2
dds <- ...

# Store results in separate variable
res <- ...
```

First few rows of results should look like this:

```{r rundeseq2, eval=T, echo=F}
# Run DESeq2
dds <- DESeq(dds)

# Store results in separate variable
res <- results(dds)

# Show results for first few genes
head(res)
```

### MA Plot

DESeq2 contains a function `plotMA` that generates an MA Plot commonly used to visualize the differential expression results. The plot shows the log2 fold changes of a given variable over the mean of normalized counts for all the samples. Points represent genes and will be colored red if the adjusted p-value is less than 0.1. Points which fall out of the window are plotted as open triangles pointing either up or down.

```{r plotMA, eval=T}
plotMA(res, ylim=c(-2,2))
```

### Significant genes

Let's add a column that tell us whether each gene shows significant log fold change between tissues. We can add a column showing the significance based on a cutoff value of padj < 0.05.

Hint: convert the results table to a data frame first.

```{r}
# Convert 'res' to data frame
res <- ...

# Add column with significant genes
res$sig <- ...

# Check first few genes after adding significance column
head(res)
```

```{r addSigGenes, eval=T, echo=F}
# Convert 'res' to data frame
res <- as.data.frame(res)

# Add column with significant genes
res$sig = res$padj < 0.05

# Check first few genes after adding significance column
head(res)
```

### Volcano plot

Volcano plots are another common way to visualize differential expression results. It shows the log2 fold change on the x-axis and the log10-transformed adjusted p-value on the y-axis (`-1*log10(padj)`), with the points colored according to whether they are significant or not. We can use `ggplot` to create a volcano plot for our results.

```{r volcanoPlot, eval=T}
# Volcano plot
ggplot(res, aes(log2FoldChange, -1*log10(padj), col=sig)) + 
  geom_point() + 
  ggtitle("Volcano plot")
```

# Gene Set Enrichment Analysis

Although we've identified differentially expressed genes, we don't yet know whether the genes have an unifying biological connection. That's where gene set enrichment analysis (GSEA) comes in. GSEA is a method to identify classes of gene that are over-represented in a large gene set. The calculated enrichment score (ES) reflects the degree to which a gene set is over-represented at the extremes (top or bottom) of the entire ranked list.

GSEA requires two inputs:
1) ranked list of genes of interest - We will be using our differentially expressed genes, ranked by their log2 fold change.
2) gene sets of biological pathways - A gene set should consist of all genes that are biological related or part of the same biological pathway. We can obtain these gene sets from the Molecular Signatures Database (MSigDB).

```{r, eval=T}
library(msigdbr)
msigdbr_show_species()
mdf <- msigdbr(species="Homo sapiens", category="H") %>%
       dplyr::select(gs_name, gene_symbol) %>%
       as.data.frame() %>%
       stats::aggregate(gene_symbol ~ gs_name, data=., c)

# Convert to list
gsets <- as.list(mdf$gene_symbol)
names(gsets) <- mdf$gs_name
```

```{r, eval=T}
names(gsets)
head(gsets$HALLMARK_APOPTOSIS)
```

```{r, eval=T}
rankedlist <- res[res$sig,'log2FoldChange']
names(rankedlist) <- row.names(res[res$sig,])
rankedlist <- rankedlist[order(rankedlist)]
head(rankedlist)
```


```{r, eval=T}
fgseaRes <- fgsea(pathways = gsets,
                  stats = rankedlist,
                  nperm = 10000)
head(fgseaRes[order(pval), ])
```

```{r, eval=T}
plotEnrichment(gsets$HALLMARK_XENOBIOTIC_METABOLISM,rankedlist)
```

***&copy; The Boys***
